File: python-impl/SymbolTable.py
from Utils import Assert
from enum import IntEnum


class SymbolScope(IntEnum):
    GLOBAL = 0,
    LOCAL = 1,
    BUILTIN_FUNCTION = 2,
    BUILTIN_VARIABLE = 3,


class Symbol:
    name: str
    isStructSymbol: bool
    scope: SymbolScope = SymbolScope.GLOBAL
    index: int
    scopeDepth: int
    isInUpScope: int

    def __init__(self, name="", scope: SymbolScope = SymbolScope.GLOBAL, index: int = 0, scopeDepth: int = 0, isStructSymbol: bool = False) -> None:
        self.name = name
        self.isStructSymbol = isStructSymbol
        self.scope = scope
        self.index = index
        self.scopeDepth = scopeDepth
        self.isInUpScope = 0


class SymbolTable:
    enclosing = None
    symbolMaps: dict[str, Symbol] = {}
    definitionCount: int
    scopeDepth: int

    def __init__(self, enclosing=None) -> None:
        self.enclosing = enclosing
        self.symbolMaps = {}
        self.definitionCount = 0
        self.scopeDepth = 0
        if self.enclosing != None:
            self.scopeDepth = self.enclosing.scopeDepth+1

    def Define(self, name: str, isStructSymbol: bool = False) -> Symbol:
        symbol = Symbol(name, SymbolScope.GLOBAL,
                        self.definitionCount, self.scopeDepth, isStructSymbol)
        if self.enclosing == None:
            symbol.scope = SymbolScope.GLOBAL
        else:
            symbol.scope = SymbolScope.LOCAL

        if self.symbolMaps.get(name, None) != None:
            Assert("Redefined variable:("+name+") in current context.")

        self.symbolMaps[name] = symbol
        self.definitionCount += 1
        return symbol

    def DefineBuiltinFunction(self, name: str, index: int) -> Symbol:
        symbol = Symbol(name, SymbolScope.BUILTIN_FUNCTION,
                        index, self.scopeDepth)
        self.symbolMaps[name] = symbol
        return symbol

    def DefineBuiltinVariable(self, name: str, index: int) -> Symbol:
        symbol = Symbol(name, SymbolScope.BUILTIN_VARIABLE,
                        index, self.scopeDepth)
        self.symbolMaps[name] = symbol
        return symbol

    def Resolve(self, name: str):
        symbol = self.symbolMaps.get(name)
        if symbol != None:
            return True, symbol
        elif self.enclosing != None:
            isFound, symbol = self.enclosing.Resolve(name)
            if isFound == False:
                return False, None
            if symbol.scope == SymbolScope.GLOBAL or symbol.scope == SymbolScope.BUILTIN_FUNCTION:
                return True, symbol

            symbol.isInUpScope = 1
            self.symbolMaps[symbol.name] = symbol
            return True, symbol

        return False, None

File: c#/libComputeDuck/SymbolTable.cs
using System.Collections.Generic;

namespace ComputeDuck
{
    public enum SymbolScope
    {
        GLOBAL,
        LOCAL,
        BUILTIN_FUNCTION,
        BUILTIN_VARIABLE
    }

    public class Symbol
    {
        public Symbol()
        {
            this.scope = SymbolScope.GLOBAL;
            this.index = 0;
            this.scopeDepth = 0;
            this.isInUpScope = 0;
            this.name = "";
        }

        public Symbol(string name,
                      SymbolScope scope,
                      int index,
                      int scopeDepth,
                      bool isStructSymbol = false)
        {
            this.name = name;
            this.isStructSymbol = isStructSymbol;
            this.scope = scope;
            this.index = index;
            this.scopeDepth = scopeDepth;
            this.isInUpScope = 0;
        }

        public string name;
        public bool isStructSymbol;
        public SymbolScope scope;
        public int index;
        public int scopeDepth;
        public int isInUpScope;
    }

    public class SymbolTable
    {
        public SymbolTable()
        {
            this.enclosing = null;
            this.symbolMaps = new Dictionary<string, Symbol>();
            this.definitionCount = 0;
            this.scopeDepth = 0;
        }

        public SymbolTable(SymbolTable enclosing)
        {
            this.enclosing = enclosing;
            this.symbolMaps = new Dictionary<string, Symbol>();
            this.definitionCount = 0;
            this.scopeDepth = enclosing.scopeDepth + 1;
        }

        public Symbol Define(string name, bool isStructSymbol = false)
        {
            Symbol symbol = new Symbol(name, SymbolScope.GLOBAL, definitionCount, scopeDepth, isStructSymbol);
            if (enclosing == null)
                symbol.scope = SymbolScope.GLOBAL;
            else
                symbol.scope = SymbolScope.LOCAL;

            if (symbolMaps.ContainsKey(name))
                Utils.Assert("Redefined variable:(" + name + ") in current context");

            symbolMaps[name] = symbol;
            definitionCount++;
            return symbol;
        }

        public Symbol DefineBuiltinFunction(string name, int index)
        {
            var symbol = new Symbol(name, SymbolScope.BUILTIN_FUNCTION, index, scopeDepth);
            symbolMaps[name] = symbol;
            return symbol;
        }

        public Symbol DefineBuiltinVariable(string name, int index)
        {
            var symbol = new Symbol(name, SymbolScope.BUILTIN_VARIABLE, index, scopeDepth);
            symbolMaps[name] = symbol;
            return symbol;
        }

        public bool Resolve(string name, out Symbol symbol)
        {
            var isFound = symbolMaps.ContainsKey(name);
            if (isFound)
            {
                symbol = symbolMaps[name];
                return true;
            }
            else if (enclosing != null)
            {
                isFound = enclosing.Resolve(name, out symbol);
                if (!isFound)
                    return false;
                if (symbol.scope == SymbolScope.GLOBAL || symbol.scope == SymbolScope.BUILTIN_FUNCTION || symbol.scope == SymbolScope.BUILTIN_VARIABLE)
                    return true;

                symbol.isInUpScope = 1;
                symbolMaps[symbol.name] = symbol;
                return true;
            }

            symbol = new Symbol();// for avoiding .net compiler error
            return false;
        }

        public SymbolTable enclosing;
        public Dictionary<string, Symbol> symbolMaps;
        public int definitionCount;
        public int scopeDepth;
    }

}